### ISSUE TITLE
Timeline for PHP 7.2 RC1 availability

### ISSUE THREAD

**Issue by jmikola** [travis-ci/travis-ci#8399]
PHP 7.2 RC1 introduced an ABI change from the previous 7.2 beta3 release, which is currently available on Travis. This change resulted in build failures for numerous projects using the `mongodb` extension (see: https://github.com/mongodb/mongo-php-driver/issues/649), as `mongodb`'s most recent [1.2.10](https://pecl.php.net/package/mongodb/1.2.10) release included fixes for PHP 7.2 RC1 compatibility (at the expense of beta3 compatibility).

Unfortunately, neither PHP nor timelib (the vendored library within PHP responsible for the ABI break) provide a proper way to differentiate beta3 from RC1.

https://github.com/php-build/php-build/pull/474 recently added support for PHP 7.2 RC1. As that is a dependency of [php-src-builder](https://github.com/travis-ci/php-src-builder/), when can we expect the various PHP environments to be updated in turn? [Tagged releases of php-build](https://github.com/php-build/php-build/releases) are quite outdated, so I'm assuming that Travis is tracking their master branch.

**Comment by BanzaiMan** [comment]
https://travis-ci.org/travis-ci/php-src-builder/builds/274240518

`7.2` is now an alias pointing to `7.2.0RC1`.

**Comment by jmikola** [comment]
Thanks for the quick resolution! https://github.com/mongodb/mongo-php-driver/pull/653 looks good and I expect various project builds will return to green soon.

**Comment by Naktibalda** [comment]
How come mongodb extension wasn't bundled with PHP 7.2 RC1, but it became bundled with RC2?

`pecl install mongodb` fails with 

`pecl/mongodb is already installed and is the same as the released version 1.2.10
install failed`
https://travis-ci.org/Codeception/Codeception/jobs/277070684#L483

**Comment by BanzaiMan** [comment]
@Naktibalda https://github.com/travis-ci/php-src-builder/commit/e4c10297910dee5a89ee7c28d3e52b955df6a261

**Comment by Naktibalda** [comment]
What is a correct workaround to make the build pass?

* `'[[ "$TRAVIS_PHP_VERSION" == "7.2" ]] || pecl install mongodb'`
* `'pecl install mongodb || true`
* something else?

**Comment by Naktibalda** [comment]
Have you actually broken builds of all projects that use `pecl install mongodb` with PHP 7.2?

My Github search skills aren't great but here is another broken build:
For example: https://travis-ci.org/doctrine/data-fixtures/jobs/274055214

**Comment by jmikola** [comment]
The commit that @BanzaiMan linked above merely changed the build script to point to the canonical Git repository (avoiding an unnecessary redirect). https://github.com/travis-ci/php-src-builder/commit/33f4aebe1e553eb54c77519d4ce4191b7fb04a96 is the actual fix that ensures `pecl install mongodb` is used for all but master and snapshot versions of PHP. Previously, all 7.x versions of PHP were installed by cloning the repository's master branch and executing:

```
phpize
./configure
make all
make install
```

The recipe for `make install` is [`install-modules`](https://github.com/php/php-src/blob/php-7.2.0RC2/Makefile.global#L42), which only copies the compiled so file into PHP's extension directory. This would be why some `.travis.yml` configurations had logic to manually add "extension=foo.so" to `php.ini` (as discussed in [Enabling preinstalled PHP extensions](https://docs.travis-ci.com/user/languages/php#Enabling-preinstalled-PHP-extensions)). `pecl install` happens to modify `php.ini` after building an extension (and may also note the installed version in some internal registry), which is what's causing the error.

I'm not sure why the swap to `pecl install` is only affecting PHP 7.2 environments, but perhaps it was the only environment rebuilt after https://github.com/travis-ci/php-src-builder/commit/33f4aebe1e553eb54c77519d4ce4191b7fb04a96 was merged. That said, you should be able to work around the issue by either ignoring the return value (as you've done above) or forcing the installation with `pecl install -f`. A force install would have the added benefit of not ignoring a possible legitimate error and also ensuring that the latest stable version is installed for each build -- obviously at the expense of additional downloads.

**Comment by BanzaiMan** [comment]
7.2.0RC1 was archived before the additional changes that affected excluded `^7` versions; so it did not run `pecl install mongodb`.

With more recent additional changes, the current logic is to use `pecl install mongodb` to install the mongodb extension on all but `master` and `*snapshot`; this, naturally, includes 7.2.0RC2.

I need your input as to what the correct behavior is, for PHP 7 builds going forward. Given @jmikola's input, I get the impression that `pecl install mongodb` was the desirable behavior going forward.

**Comment by jmikola** [comment]
@BanzaiMan: I believe the general problem is that the scripts in [/default/bin](https://github.com/travis-ci/php-src-builder/tree/default/bin) use various installation methods. Consider:

 * [amqp](https://github.com/travis-ci/php-src-builder/blob/default/bin/compile-extension-amqp) is always compiled manually from a git clone. Its `amqp.so` file is copied to PHP's extension directory via `make install` but PECL has no record of the extension being installed.
 * [apcu](https://github.com/travis-ci/php-src-builder/blob/default/bin/compile-extension-apcu) is installed via `pecl install`, so PECL's registry will note that the package is installed. Depending on whether `php_ini` is defined for PECL, `extension=apcu.so` may also be added to the INI configuration (i.e. it will be enabled and loaded by default).
 * [memcached](https://github.com/travis-ci/php-src-builder/blob/default/bin/compile-extension-memcached) is always compiled manually. Depending on the environment, source may be obtained from either git clone or via `pecl download`; however, since `pecl install` is never executed PECL's registry will have no record of the extension.
 * [mongodb](https://github.com/travis-ci/php-src-builder/blob/default/bin/compile-extension-mongodb) is compiled manually from a git clone for `master` and `snapshot` environments, but is installed via `pecl install` otherwise.

The steps in [Enabling preinstalled PHP extensions](https://docs.travis-ci.com/user/languages/php#Enabling-preinstalled-PHP-extensions) only make sense for extensions whose script ends with `make install`, since that copies the `extension.so` file to PHP's directory but doesn't modify PHP's INI configuration to actually load the extension.

Based on [this build log](https://travis-ci.org/mongodb/mongo-php-library/jobs/275978364#L1429) from [mongodb/mongo-php-library](https://github.com/mongodb/mongo-php-library), which runs `pecl install mongodb` on Travis, PECL does have `php_ini` defined and will actually modify PHP's INI files to load an installed extension:

```
Build process completed successfully
Installing '/home/travis/.phpenv/versions/7.0.22/lib/php/extensions/no-debug-zts-20151012/mongodb.so'
install ok: channel://pecl.php.net/mongodb-1.3.0RC1
Extension mongodb enabled in php.ini
````

Assuming PECL configurations are consistent across the various PHP environments, we can assume that anything using `pecl install` will enable an extension, too. In that case, [Enabling preinstalled PHP extensions](https://docs.travis-ci.com/user/languages/php#Enabling-preinstalled-PHP-extensions) will not be relevant, but [Disabling preinstalled PHP extensions](https://docs.travis-ci.com/user/languages/php#Disabling-preinstalled-PHP-extensions) may still apply.

**Comment by BanzaiMan** [comment]
@jmikola Different installation methods are used for different extensions because, at the time I was developing the tool, most extensions did not work with PHP 7. My understanding was that installation from source was functional equivalent.

If this is not the case, then it seems far better to me to not make *any* sort of promise about what extensions are pre-installed.

**Comment by Naktibalda** [comment]
It looks like it is no longer necessary to add `-f`.
Since Friday `pecl install mongodb` works successfully on PHP 7.2 (we had no builds on Thursday),
But it fails on PHP 5.4 because mongodb 1.3.0 (released on 19th) raised the minimum supported version to PHP 5.5.
Isn't pecl install supposed to install a compatible version.

https://travis-ci.org/Codeception/Codeception/jobs/278594475#L493

```
$ pecl install mongodb
WARNING: channel "pecl.php.net" has updated its protocols, use "pecl channel-update pecl.php.net" to update
pecl/mongodb requires PHP (version >= 5.5.0, version <= 7.99.99), installed version is 5.4.45
No valid packages found
install failed

**Comment by jmikola** [comment]
> If this is not the case, then it seems far better to me to not make any sort of promise about what extensions are pre-installed.

That sounds fair. I've personally never relied on pre-installed extensions.

The alternative would be attempting to swap all of the build scripts over to `pecl install`, which may be feasible since most should have releases that support PHP 7. However, if packages are no longer compatible with older PHP versions, `pecl install` may need to specify an exact version (as is the case below).

> Isn't pecl install supposed to install a compatible version.

I would not be surprised if PECL lacked support for selecting an older package version to match the PHP environment. That is definitely functionality we can expect from Composer, but PECL is much more naive. You may need to specifically refer to `mongodb-1.2.11` for your PHP 5.4 builds.

**Comment by Naktibalda** [comment]
And it flipped again, `pecl install mongodb` fails on 7.1 and 7.2 with

`pecl/mongodb is already installed and is the same as the released version 1.3.0`

**Comment by xtoolkit** [comment]
`$ pecl install mongodb`
pecl/mongodb is already installed and is the same as the released version 1.5.2
install failed
and again
